app/Http/Controllers/BookController.php

```diff
@@ -24,4 +28,73 @@ public function detail($slug)
         }
 
     }
+
+    public function showByCategory($slug)
+    {
+        try {
+            $category = Category::where('slug', $slug)->firstOrFail();
+            $books = $category->books()->paginate(config('const.take'));
+
+            return view('user.pages.booksview', compact('books'));
+        } catch (ModelNotFoundException $e) {
+            response()->view('errors.404_user_not_found', [], 404);
+        }
+    }
+
+    public function showByPublisher($slug)
+    {
+        try {
+            $publisher = Publisher::where('slug', $slug)->firstOrFail();
+            $books = $publisher->books()->paginate(config('const.take'));
+
+            return view('user.pages.booksview', compact('books'));
+        } catch (ModelNotFoundException $e) {
+            response()->view('errors.404_user_not_found', [], 404);
+        }
+    }
+
+    public function showByAuthor($slug)
+    {
+        try {
+            $author = User::where('user_slug', $slug)->firstOrFail();
+            $books = $author->books()->paginate(config('const.take'));
+
+            return view('user.pages.booksview', compact('books'));
+        } catch (ModelNotFoundException $e) {
+            response()->view('errors.404_user_not_found', [], 404);
+        }
+    }
+
+    public function searchByCategory(Request $request)
+    {
+        $keywords = trim($request->input('keywords'));
```

Comment:
```suggestion
        $keywords = trim($request->'keywords');
```

Reply:
vanga

Reply:
khong duoc chi oi

Reply:
```suggestion
        $keywords = trim($request->keywords);
```
nhầm 1 tý, ko biết bỏ cái dấu '' đi à. chẹp

Reply:
hi em copy cho nhanh i mak, de em sua a